pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -928,6 +943,27 @@ public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
         }
     }
 
+    private int removeExpiredMessagesFromQueue(Set<MessageIdImpl> messageIds) {
+        int messagesFromQueue = 0;
+        for (Message message : incomingMessages) {
```

Comment:
We could use a LinkedBlockingQueue though this would generate more garbage, though this should only affect the client, not the broker.
I'm not aware of an implementation of linked queues that pools objects to prevent garbage generation.